build/lib/src/builder/build_step_impl.dart

```diff
@@ -89,14 +90,14 @@ class BuildStepImpl implements BuildStep {
   Future writeAsString(AssetId id, FutureOr<String> content,
       {Encoding encoding: UTF8}) {
     _checkOutput(id);
-    var done = _futureOrWrite(
+    var done = _futureOrWrite<String>(
         content, (c) => _writer.writeAsString(id, c, encoding: encoding));
     _outputsCompleted = _outputsCompleted.then((_) => done);
     return done;
   }
 
   Future _futureOrWrite<T>(FutureOr<T> content, Future write(T content)) =>
-      (content is Future<T>) ? content.then(write) : write(content);
+      (content is Future<T>) ? content.then(write) : write(content as T);
```

Comment:
I'm hoping the enhanced type promotion stuff Bob was looking at will help here. This is in line with the places where an `is` check lets you treat the variable as if it had the cast, but an `is!` doesn't, and the `else` clause doesn't get type promotion from either.